clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -511,6 +548,10 @@ public boolean hasUndrained() {
                             ProducerBatch first = deque.peekFirst();
                             if (first != null) {
                                 boolean backoff = first.attempts() > 0 && first.waitedTimeMs(now) < retryBackoffMs;
+                                // We determine the earliest time any batch may expire. We use
+                                // this value later on to determine the max time we can sleep in poll.
+                                earliestDeliveryTimeoutMs = Math.min(earliestDeliveryTimeoutMs, first.createdMs + deliveryTimeoutMs);
```

Comment:
The logic for updating this field seems to assume that the batch at the front of the deque will always be the next to expire, but I'm not sure that is true in the case of retries.